core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala

```diff
@@ -65,13 +66,121 @@ class JDBCSourceAsColumnarStore(override val connProperties: ConnectionPropertie
+      (implicit conn: Option[Connection] = None): Unit = {
     // noinspection RedundantDefaultArgument
-    tryExecute(tableName, doInsert(tableName, batch, batchId,
-      getPartitionID(tableName, partitionId), maxDeltaRows),
-      closeOnSuccess = true, onExecutor = true)
+    tryExecute(tableName, closeOnSuccessOrFailure = false, onExecutor = true)(doInsert(tableName, batch, batchId,
+      getPartitionID(conn, tableName, partitionId), maxDeltaRows))(implicitly, conn)
   }
 
+  // begin should decide the connection which will be used by insert/commit/rollback
+  def beginTx(): Array[_ <: Object] = {
+
+    implicit val conn = self.getConnection(tableName, onExecutor = true)
+
+    assert(!conn.isClosed)
+    tryExecute(tableName, closeOnSuccessOrFailure = false, onExecutor = true) {
+      (conn: Connection) => {
+        connectionType match {
+          case ConnectionType.Embedded =>
+            val txMgr = Misc.getGemFireCache().getCacheTransactionManager()
+            if (txMgr.getTXState() == null) {
+              txMgr.begin(com.gemstone.gemfire.cache.IsolationLevel.SNAPSHOT, null)
+              Array(conn, txMgr.getTransactionId().stringFormat())
+            } else {
+              Array(conn, null)
+            }
+          case _ =>
+            val txId = SparkShellRDDHelper.snapshotTxIdForWrite.get
+            if (txId == null) {
+              logDebug("Going to start the transaction on server ")
+              val startAndGetSnapshotTXId = conn.prepareCall(s"call sys.START_SNAPSHOT_TXID (?)")
+              startAndGetSnapshotTXId.registerOutParameter(1, java.sql.Types.VARCHAR)
+              startAndGetSnapshotTXId.execute()
+              val txid: String = startAndGetSnapshotTXId.getString(1)
+              startAndGetSnapshotTXId.close()
+              SparkShellRDDHelper.snapshotTxIdForWrite.set(txid)
+              logDebug(s"The snapshot tx id is ${txid} and tablename is ${tableName}")
+              Array(conn, txid)
+            } else {
+              // it should always be not null.
+              if (!txId.equals("null")) {
+                val statement = conn.createStatement()
+                statement.execute(
+                  s"call sys.USE_SNAPSHOT_TXID('$txId')")
+              }
+              Array(conn, null)
+            }
+        }
+      }
+    } (implicitly, Some(conn))
+  }
+
+  def commitTx(txId: String)(implicit conn: Option[Connection]): Unit = {
+    tryExecute(tableName, closeOnSuccessOrFailure = true, onExecutor = true) {
+      (conn: Connection) => {
+        connectionType match {
+          case ConnectionType.Embedded =>
+            Misc.getGemFireCache().getCacheTransactionManager().commit()
+          case _ =>
+            logDebug(s"Going to commit ${txId} the transaction on server ")
+            val ps = conn.prepareStatement(s"call sys.COMMIT_SNAPSHOT_TXID(?)")
+            ps.setString(1, if (txId == null) "null" else txId)
```

Comment:
This would not happen mostly in case of insert. But this used to happen in read in case of join. Will try to test it.
However, it will be no-op on server if txId is null.